src/language.rs

```diff
@@ -129,6 +142,20 @@ pub trait SynthLanguage: egg::Language + Send + Sync + Display + FromOp + 'stati
         }
     }
 
+    fn is_allowed_rewrite(lhs: &Pattern<Self>, rhs: &Pattern<Self>) -> bool {
+        let pattern_is_extractable = |pat: &Pattern<Self>| {
+            pat.ast.as_ref().iter().all(|n| match n {
+                ENodeOrVar::ENode(n) => n.is_extractable(),
+                ENodeOrVar::Var(_) => true,
+            })
+        };
+        pattern_is_extractable(lhs) && pattern_is_extractable(rhs)
+    }
+
+    fn is_extractable(&self) -> bool {
```

Comment:
assuming rule-lifting related parts of the code are not for review for now.

Reply:
yeah, not quite